src/tmpfiles/tmpfiles.c

```diff
@@ -1288,45 +1287,113 @@ static int create_file(Item *i, const char *path) {
         }
 
         if (fd < 0) {
-                if (i->type == CREATE_FILE && errno == EEXIST) {
-                        log_debug_errno(errno, "Not writing to pre-existing file \"%s\": %m", path);
-                        goto done;
-                }
+                /* Even on a read-only filesystem, open(2) returns EEXIST if the
+                 * file already exists. It returns EROFS only if it needs to
+                 * create the file. */
+                if (errno != EEXIST)
+                        return log_error_errno(errno, "Failed to create file %s: %m", path);
+
+                /* Re-open the file. At that point it must exist since open(2)
+                 * failed with EEXIST. We still need to check if the perms/mode
+                 * need to be changed. For read-only filesystems, we let
+                 * fd_set_perms() report the error if the perms need to be
+                 * modified. */
+                fd = open(path, O_NOFOLLOW|O_CLOEXEC|O_PATH, i->mode);
+                if (fd < 0)
+                        return log_error_errno(errno, "Failed to re-open file %s: %m", path);
 
-                r = -errno;
-                if (!i->argument && errno == EROFS && stat(path, &st) == 0 &&
-                    (i->type == CREATE_FILE || st.st_size == 0))
-                        goto check_mode;
+                if (fstat(fd, &st) < 0)
+                        return log_error_errno(errno, "stat(%s) failed: %m", path);
 
-                return log_error_errno(r, "Failed to create file %s: %m", path);
+                if (!S_ISREG(st.st_mode))
+                        return log_error_errno(EEXIST, "%s exists and is not a regular file.", path);
+
+                goto check_perms;
```

Comment:
Maybe get rid of the `goto` and just put the code that is jumped over in an `else { }` branch?

Reply:
ok

Reply:
also, here too we'll stat the file twice in the end, I think it would be much nicer to pass "struct stat" if we have it, like in this case, and only query it in fd_set_perms() if we don't.